In [1]:
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torch.utils.data import DataLoader
import torchvision
import torchvision.transforms as transforms
from dataclasses import dataclass
import matplotlib.pyplot as plt
import numpy as np
from tqdm import tqdm
import math

In [ ]:
@dataclass
class SiglipVisionConfig:
    num_hidden_layers: int=6
    num_channels: int=3          #channels in input image :3-RGB
    image_size: int=32           #CIFAR10 dataset images are 32x32
    patch_size: int=4            #4x4 patches for 32x32 :64 patches
    num_attention_heads: int=8   #number of attention heads
    hidden_size: int=384
    intermediate_size: int=1536
    num_classes: int=10        #since cifar10 has 10 classes
    layer_norm_eps: float=1e-6
    attention_dropout: float=0.1
    dropout: float=0.1

def get_cifar10_transforms():

    train_transform = transforms.Compose([
        transforms.RandomHorizontalFlip(p=0.5),
        transforms.RandomCrop(32, padding=4),
        transforms.ToTensor(),               #converts the images to a tensor
        transforms.Normalize(mean=[0.4914, 0.4822, 0.4465], std=[0.247, 0.243, 0.261])  #normalizes the image tensor
    ])

    test_transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.4914, 0.4822, 0.4465], std=[0.247, 0.243, 0.261])
    ])
    return train_transform, test_transform


In [ ]:
class SiglipVisionEmbeddings(nn.Module):
    def __init__(self, config: SiglipVisionConfig):
        super().__init__()
        self.config = config

        #basic parameters -
        self.num_channels = config.num_channels
        self.embed_dim = config.hidden_size #each image patch is going to be turned into a vector of 384 dimensions
        self.image_size = config.image_size #image size is 32x32 pixels
        self.patch_size = config.patch_size #each patch has a size of 4x4 pixels

        #convolution used to create patch embeddings
        self.patch_embedding = nn.Conv2d(
            in_channels=self.num_channels,
            out_channels=self.embed_dim,
            kernel_size=self.patch_size,
            stride=self.patch_size,
            padding='valid'
        )

        #calculating the number of patches were going to get-
        self.num_patches = (self.image_size//self.patch_size)**2   #32/4=8x8=64 patches
        self.num_positions = self.num_patches + 1                     #+1 for CLS token

        self.cls_token = nn.Parameter(torch.zeros(1, 1, self.embed_dim))
        #creating the positional embedding layer -> creates a lookup table of size num_patches x embed_dim
        self.position_embedding = nn.Embedding(self.num_positions, self.embed_dim)

        self.register_buffer(
            #storing position indices -
            "position_ids",
            torch.arange(self.num_positions).expand(1, -1),
            persistent=False
        )
        self.dropout = nn.Dropout(config.dropout)

    def forward(self, pixel_values: torch.FloatTensor) -> torch.Tensor:
        #getting batch size and image dimensions -
        B,C,H,W= pixel_values.shape # assigning the pixel values to Batch, Channel, Height and Width

        patch_embeds = self.patch_embedding(pixel_values)
        #flattening and reshaping patches
        embeddings=patch_embeds.flatten(2).transpose(1, 2)  # B, num_patches, embed_dim

        #adding CLS token
        cls_tokens=self.cls_token.expand(B, -1, -1)
        embeddings=torch.cat((cls_tokens, embeddings),dim=1)

        #adding the position embeddings to patch embeddings
        embeddings=embeddings+self.position_embedding(self.position_ids)
        return self.dropout(embeddings)

In [ ]:
# MLP LAYER

class SiglipMLP(nn.Module):
    def __init__(self, config: SiglipVisionConfig):
        super().__init__()
        self.config = config
        #fully connected layer 1
        self.fc1 = nn.Linear(config.hidden_size, config.intermediate_size)
        #fully connected layer 2
        self.fc2 = nn.Linear(config.intermediate_size, config.hidden_size)
        #using dropout for regularization -
        self.dropout = nn.Dropout(config.dropout)

    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
        hidden_states = self.fc1(hidden_states)
        #applying the non linearity activation function -
        hidden_states = F.gelu(hidden_states)
        hidden_states = self.dropout(hidden_states)
        hidden_states = self.fc2(hidden_states)
        return hidden_states

In [ ]:
# ATTENTION LAYER (part of Encoder Layer)

class SiglipAttention(nn.Module):
    def __init__(self, config: SiglipVisionConfig):
        super().__init__()
        self.config = config
        self.embed_dim = config.hidden_size
        self.num_heads = config.num_attention_heads
        self.head_dim = self.embed_dim // self.num_heads # dimensions per head
        self.dropout = config.attention_dropout

        #initialising three linear transformations for key, query and value
        self.q_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=True)
        self.k_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=True)
        self.v_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=True)

        self.out_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=True)  #this gives the output projection

    def forward(self, hidden_states):
        B, T, C = hidden_states.shape # B=batch,T= tokens,C=channels

        q_states = self.q_proj(hidden_states).view(B,T,self.num_heads, C // self.num_heads).transpose(1, 2)
        k_states = self.k_proj(hidden_states).view(B,T,self.num_heads, C // self.num_heads).transpose(1, 2)
        v_states = self.v_proj(hidden_states).view(B,T,self.num_heads, C // self.num_heads).transpose(1, 2)

        #scaled dot product attention -
        attn_weights = (q_states @ k_states.transpose(-2, -1)) / math.sqrt(self.head_dim)
        attn_weights = F.softmax(attn_weights, dim=-1)

        attn_weights = F.dropout(attn_weights, p=self.dropout, training=self.training)

        # multiply attention with values -
        attn_outs=attn_weights@v_states

        attn_outs = attn_outs.transpose(1,2).reshape(B,T,C)
        return self.out_proj(attn_outs)

In [ ]:
# ENCODER LAYER

class SiglipEncoderLayer(nn.Module):
    def __init__(self, config: SiglipVisionConfig):
        super().__init__()
        self.embed_dim = config.hidden_size
        self.self_attn = SiglipAttention(config)
        self.layer_norm1 = nn.LayerNorm(self.embed_dim, eps=config.layer_norm_eps)
        self.mlp = SiglipMLP(config)
        self.layer_norm2 = nn.LayerNorm(self.embed_dim, eps=config.layer_norm_eps)

    def forward(self, hidden_states):
        # first residual block
        residual = hidden_states
        hidden_states = self.layer_norm1(hidden_states)
        hidden_states = self.self_attn(hidden_states)
        hidden_states += residual

        # second residual block
        residual = hidden_states
        hidden_states = self.layer_norm2(hidden_states)
        hidden_states = self.mlp(hidden_states)
        hidden_states += residual
        return hidden_states

In [ ]:
# FULL ENCODER

class SiglipEncoder(nn.Module):
    def __init__(self, config: SiglipVisionConfig):
        super().__init__()
        self.config = config
        # stacking multiple encoder layers -
        self.layers = nn.ModuleList([SiglipEncoderLayer(config) for _ in range(config.num_hidden_layers)])

    def forward(self, hidden_states):
        for layer in self.layers:
            hidden_states = layer(hidden_states)
        return hidden_states

In [ ]:
# complete VISION TRANSFORMER for CIFAR10 classification

class CIFAR10VisionTransformer(nn.Module):
    def __init__(self, config: SiglipVisionConfig):
        super().__init__()
        self.config = config
        self.embeddings = SiglipVisionEmbeddings(config)
        self.encoder = SiglipEncoder(config)
        self.layer_norm = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)

        # classification head
        self.classifier = nn.Linear(config.hidden_size, config.num_classes)

    def forward(self, pixel_values):
        hidden_states = self.embeddings(pixel_values)
        hidden_states = self.encoder(hidden_states)
        hidden_states = self.layer_norm(hidden_states)

        # Use CLS token (first token) for classification
        cls_token = hidden_states[:, 0]
        logits = self.classifier(cls_token)
        return logits



In [ ]:
# DATA LOADING AND TRAINING

def main():
    # loading CIFAR10 dataset
    train_transform, test_transform = get_cifar10_transforms()

    trainset = torchvision.datasets.CIFAR10(
        root='./data',
        train=True,
        download=True,
        transform=train_transform
    )
    testset = torchvision.datasets.CIFAR10(
        root='./data',
        train=False,
        download=True,
        transform=test_transform
    )

    trainloader = DataLoader(trainset, batch_size=128, shuffle=True, num_workers=4, pin_memory=True)
    testloader = DataLoader(testset, batch_size=128, shuffle=False, num_workers=4, pin_memory=True)

    # CIFAR10 classes names
    classes = ('plane', 'car', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse', 'ship', 'truck')

    # Model, loss, optimizer
    config = SiglipVisionConfig()
    model = CIFAR10VisionTransformer(config).cuda()
    criterion = nn.CrossEntropyLoss() # loss function
    optimizer = optim.AdamW(model.parameters(), lr=3e-4, weight_decay=0.05)
    scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=50)



    # Training loop
    num_epochs = 50
    best_acc = 0.0

    for epoch in range(num_epochs):
        # Training
        model.train()
        train_loss = 0.0
        train_correct = 0
        train_total = 0

        pbar = tqdm(trainloader, desc=f'Epoch {epoch+1}/{num_epochs}')
        for batch_idx, (inputs, targets) in enumerate(pbar):
            inputs, targets = inputs.cuda(), targets.cuda()

            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            loss.backward()

            # gradient clipping
            torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
            optimizer.step()

            train_loss += loss.item()
            _, predicted = outputs.max(1)
            train_total += targets.size(0)
            train_correct += predicted.eq(targets).sum().item()

            pbar.set_postfix({
                'Loss': f'{loss.item():.3f}',
                'Acc': f'{100.*train_correct/train_total:.1f}%'
            })

        scheduler.step()




        # Evaluation
        model.eval()
        test_loss = 0.0
        test_correct = 0
        test_total = 0

        with torch.no_grad():
            for inputs, targets in testloader:
                inputs, targets = inputs.cuda(), targets.cuda()
                outputs = model(inputs)
                loss = criterion(outputs, targets)

                test_loss += loss.item()
                _, predicted = outputs.max(1)
                test_total += targets.size(0)
                test_correct += predicted.eq(targets).sum().item()

        train_acc = 100. * train_correct / train_total
        test_acc = 100. * test_correct / test_total


        print(f'Epoch {epoch+1}: Train Acc: {train_acc:.2f}%, Test Acc: {test_acc:.2f}%')




        if test_acc > best_acc:
            best_acc = test_acc
            torch.save(model.state_dict(), 'best_cifar10_vit.pth')

    print(f'Best Test Accuracy: {best_acc:.2f}%')

if __name__ == "__main__":
    main()


100%|██████████| 170M/170M [00:03<00:00, 46.1MB/s]
/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:627: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(
Epoch 1/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=1.561, Acc=34.3%]


Epoch 1: Train Acc: 34.33%, Test Acc: 48.09%


Epoch 2/50: 100%|██████████| 391/391 [01:18<00:00,  4.96it/s, Loss=1.294, Acc=45.8%]


Epoch 2: Train Acc: 45.77%, Test Acc: 51.37%


Epoch 3/50: 100%|██████████| 391/391 [01:19<00:00,  4.94it/s, Loss=1.420, Acc=50.2%]


Epoch 3: Train Acc: 50.21%, Test Acc: 55.09%


Epoch 4/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=1.052, Acc=53.2%]


Epoch 4: Train Acc: 53.16%, Test Acc: 56.31%


Epoch 5/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=1.164, Acc=55.9%]


Epoch 5: Train Acc: 55.87%, Test Acc: 60.34%


Epoch 6/50: 100%|██████████| 391/391 [01:19<00:00,  4.94it/s, Loss=1.120, Acc=58.2%]


Epoch 6: Train Acc: 58.23%, Test Acc: 60.31%


Epoch 7/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.983, Acc=60.3%]


Epoch 7: Train Acc: 60.35%, Test Acc: 63.33%


Epoch 8/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.960, Acc=62.3%]


Epoch 8: Train Acc: 62.27%, Test Acc: 64.14%


Epoch 9/50: 100%|██████████| 391/391 [01:19<00:00,  4.94it/s, Loss=0.845, Acc=64.2%]


Epoch 9: Train Acc: 64.22%, Test Acc: 65.68%


Epoch 10/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.982, Acc=65.9%]


Epoch 10: Train Acc: 65.90%, Test Acc: 67.25%


Epoch 11/50: 100%|██████████| 391/391 [01:19<00:00,  4.95it/s, Loss=0.954, Acc=67.6%]


Epoch 11: Train Acc: 67.63%, Test Acc: 69.18%


Epoch 12/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.727, Acc=68.9%]


Epoch 12: Train Acc: 68.87%, Test Acc: 68.85%


Epoch 13/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.767, Acc=69.9%]


Epoch 13: Train Acc: 69.87%, Test Acc: 70.93%


Epoch 14/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.824, Acc=71.4%]


Epoch 14: Train Acc: 71.41%, Test Acc: 71.16%


Epoch 15/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.628, Acc=72.4%]


Epoch 15: Train Acc: 72.35%, Test Acc: 72.36%


Epoch 16/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.789, Acc=73.4%]


Epoch 16: Train Acc: 73.41%, Test Acc: 73.37%


Epoch 17/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.537, Acc=74.7%]


Epoch 17: Train Acc: 74.68%, Test Acc: 73.91%


Epoch 18/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.710, Acc=75.6%]


Epoch 18: Train Acc: 75.61%, Test Acc: 74.72%


Epoch 19/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.714, Acc=76.4%]


Epoch 19: Train Acc: 76.35%, Test Acc: 74.57%


Epoch 20/50: 100%|██████████| 391/391 [01:19<00:00,  4.94it/s, Loss=0.576, Acc=77.4%]


Epoch 20: Train Acc: 77.39%, Test Acc: 75.55%


Epoch 21/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.628, Acc=78.5%]


Epoch 21: Train Acc: 78.46%, Test Acc: 75.50%


Epoch 22/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.642, Acc=79.4%]


Epoch 22: Train Acc: 79.35%, Test Acc: 75.89%


Epoch 23/50: 100%|██████████| 391/391 [01:19<00:00,  4.94it/s, Loss=0.619, Acc=80.1%]


Epoch 23: Train Acc: 80.10%, Test Acc: 76.24%


Epoch 24/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.758, Acc=81.1%]


Epoch 24: Train Acc: 81.13%, Test Acc: 76.93%


Epoch 25/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.562, Acc=81.9%]


Epoch 25: Train Acc: 81.91%, Test Acc: 77.03%


Epoch 26/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.490, Acc=82.7%]


Epoch 26: Train Acc: 82.73%, Test Acc: 77.03%


Epoch 27/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.611, Acc=83.6%]


Epoch 27: Train Acc: 83.61%, Test Acc: 77.12%


Epoch 28/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.492, Acc=84.5%]


Epoch 28: Train Acc: 84.52%, Test Acc: 77.52%


Epoch 29/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.317, Acc=85.4%]


Epoch 29: Train Acc: 85.41%, Test Acc: 77.48%


Epoch 30/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.359, Acc=86.2%]


Epoch 30: Train Acc: 86.17%, Test Acc: 78.25%


Epoch 31/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.451, Acc=86.9%]


Epoch 31: Train Acc: 86.91%, Test Acc: 78.41%


Epoch 32/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.341, Acc=87.9%]


Epoch 32: Train Acc: 87.91%, Test Acc: 78.68%


Epoch 33/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.232, Acc=88.6%]


Epoch 33: Train Acc: 88.61%, Test Acc: 78.83%


Epoch 34/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.343, Acc=89.2%]


Epoch 34: Train Acc: 89.23%, Test Acc: 78.96%


Epoch 35/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.245, Acc=90.1%]


Epoch 35: Train Acc: 90.07%, Test Acc: 78.84%


Epoch 36/50: 100%|██████████| 391/391 [01:19<00:00,  4.94it/s, Loss=0.241, Acc=90.6%]


Epoch 36: Train Acc: 90.61%, Test Acc: 78.19%


Epoch 37/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.223, Acc=91.1%]


Epoch 37: Train Acc: 91.08%, Test Acc: 78.80%


Epoch 38/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.236, Acc=91.8%]


Epoch 38: Train Acc: 91.84%, Test Acc: 79.09%


Epoch 39/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.133, Acc=92.4%]


Epoch 39: Train Acc: 92.38%, Test Acc: 78.86%


Epoch 40/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.146, Acc=92.8%]


Epoch 40: Train Acc: 92.81%, Test Acc: 79.43%


Epoch 41/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.116, Acc=93.2%]


Epoch 41: Train Acc: 93.18%, Test Acc: 79.35%


Epoch 42/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.118, Acc=93.7%]


Epoch 42: Train Acc: 93.73%, Test Acc: 78.86%


Epoch 43/50: 100%|██████████| 391/391 [01:19<00:00,  4.90it/s, Loss=0.177, Acc=94.0%]


Epoch 43: Train Acc: 93.95%, Test Acc: 79.31%


Epoch 44/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.171, Acc=94.2%]


Epoch 44: Train Acc: 94.16%, Test Acc: 79.55%


Epoch 45/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.103, Acc=94.3%]


Epoch 45: Train Acc: 94.30%, Test Acc: 79.24%


Epoch 46/50: 100%|██████████| 391/391 [01:19<00:00,  4.92it/s, Loss=0.126, Acc=94.5%]


Epoch 46: Train Acc: 94.52%, Test Acc: 79.40%


Epoch 47/50: 100%|██████████| 391/391 [01:19<00:00,  4.94it/s, Loss=0.290, Acc=94.8%]


Epoch 47: Train Acc: 94.81%, Test Acc: 79.47%


Epoch 48/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.096, Acc=94.8%]


Epoch 48: Train Acc: 94.76%, Test Acc: 79.58%


Epoch 49/50: 100%|██████████| 391/391 [01:19<00:00,  4.93it/s, Loss=0.183, Acc=94.7%]


Epoch 49: Train Acc: 94.72%, Test Acc: 79.61%


Epoch 50/50: 100%|██████████| 391/391 [01:19<00:00,  4.91it/s, Loss=0.138, Acc=95.0%]


Epoch 50: Train Acc: 95.00%, Test Acc: 79.57%
Best Test Accuracy: 79.61%


The Vision Transformer implemented from scratch was successfully trained on the CIFAR-10 dataset.

The model achieved a ***Training accuracy*** of ***~95%*** and a ***Best Test accuracy*** of ***~79%***.

This indicates that the model exhibits a noticeable gap between training and testing performance. This behavior is expected for models trained on small datasets such as CIFAR-10.

